### PR TITLE
[DO NOT MERGE]Add AccessibleFormatRequest route, controller and views

### PR DESCRIPTION
**Pull request by edwardkerry**
We are improving the experience of requesting a document in an accessible format. Currently users are presented with a link to an email template for them to send to the department themselves. We are moving this to a form which will collect their details and format needs and instead send an email via Notify to the department.

In practice, this form will be accessed via a link in Whitehall which will POST with the content_id of the parent document and the attachment_id of the inaccessible attachment.

The Controller reads a series of questions from a supplied YAML file, which are then presented in order. Questions can be presented as radio-buttons, a single free-text input or a series of text input boxes.

On submitting a question, a hidden "question_number" parameter is passed along with the entered parameters which is used to increment the next question to be displayed.

In order to persist entered information, a hidden form element is used to pass previous parameters with subsequent requests.

There is a concept of an optional question which is displayed only if certain conditions are met. If the required parameter and key and value are present, an optional question will display. If they are not, the next question will instead render.

On the final question, all gathered parameters are used to creates an AccessibleFormatRequest model. If valid, on save this model manipulates the various passed parameters and presents them to Notify to generate an email.

### CODE REVIEW

File: app/lib/accessible_format_request/questions.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    trigger-value: format-other
+  - key: contact-details
+    type: multi-text
+    text: In order for us to reach you about your request
+    caption: Enter your contact details
+    optional: false
+    options:
+      - label: First name
+        value: first_name
+      - label: Second name
```
Comment (maxgds) [inline app/lib/accessible_format_request/questions.yaml:34]: Might need to run this by content/design but is Second name right here? My second name is my middle name, could we use "last name"?

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 2]
```diff
@@ -83,4 +80,18 @@ def display_optional_question?
...
+  end
+
+  def multiple_keys
+    multi_key_questions = questions.select do |question|
+      question["type"] == "multi-text"
+    end
+
+    multi_key_questions.map do |question|
+      question["options"].map { |op| :"#{op["value"]}" }
+    end
```
Comment (ChrisBAshton) [inline app/controllers/contact/govuk/accessible_format_requests_controller.rb:95]: Rather than having to differentiate between 'single key' questions and 'multi key questions', I think you could standardise your YAML so that they all have the same configuration. Example:

```yaml
questions:
  - group: set-format
    title: Which accessible format would you like for
    caption: Please choose one
    optional: false
    inputs:
      - type: radio
        key: format
        options:
          - text: Audio format
            value: format-audio
          - text: Braille
            value: format-braile
          - text: British Sign Language
            value: format-bsl
          - text: Large print version
            value: format-large-print
          - text: Not seeing what you're looking for?
            value: format-other
  - group: custom-format
    title: What accessible format do you require?
    caption: Please enter your details here
    optional: true
    trigger-key: set-format
    trigger-value: format-other
    inputs:
      - type: text
        key: custom_format
  - group: contact-details
      title: In order for us to reach you about your request
      caption: Enter your contact details
      optional: false
      inputs:
        - type: text
          label: First name
          key: first_name
        - type: text
          label: Second name
          key: second_name
        - type: text
          label: Email address
          key: email
```

Then your `permitted_params` would be something like `questions.map(&:inputs).pluck(:key).flatten`. But you may want to consider combining the input key with the 'group' to avoid collisions, e.g. `first_name` => `contact-details::first_name`.

File: app/lib/accessible_format_request/questions.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+---
+questions:
+  - key: set-format
+    type: radio
+    text: Which accessible format would you like for
```
Comment (ChrisBAshton) [inline app/lib/accessible_format_request/questions.yaml:5]: This reads a bit strange to me - should it be this?

```suggestion
    text: Which accessible format would you like?
```

**Review by ChrisBAshton — COMMENTED** [review]
Hi Ed 👋 this looks like a sizeable piece of work. Well done for getting the multi-page form / persistence working nicely!

I've made a few suggestions below, some of which would require a bit of re-architecture. Apologies if that throws a spanner in the works! Let me know if some of the suggestions are unworkable, you know a lot more about what is trying to be achieved than I do. 👍

File: app/lib/accessible_format_request/questions.yaml (hunk 3 above)
Comment (edwardkerry) [reply app/lib/accessible_format_request/questions.yaml:5]: Eventually the idea is for the document name to be optionally passed in as an additional param, to be `Which accessible format would you like for #{document_name}?` - with a placeholder available if the param is not present.

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      end
+    end
+
+    context "on the 2nd page" do
+      before { get :new, params: { page: "2", 'set-format': example_param } }
+      it "renders the 2nd question text" do
+        expect(response.body).to include(format_request_questions[1]["text"])
+      end
+
+      it "displays a 'Next' button " do
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:24]: thank you!

File: app/views/contact/govuk/accessible_format_requests/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  <% if current_question[:type] == "radio" %>
+    <%= render "govuk_publishing_components/components/radio", {
+      name: current_question[:key],
+      heading: current_question[:caption],
+      items: current_question[:options]
+    } %>
+  <% end %>
+
+  <% if current_question[:type] == "text"%>
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/views/contact/govuk/accessible_format_requests/new.html.erb]: :+1:

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def questions
+    @questions ||= question_file["questions"]
+  end
+
+  def current_question
+    questions[page - 1]
+  end
+  helper_method :current_question
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb]: I went to-and-fro with this, but think I agree with you here!

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  def create; end
+
+private
+
+  def question_file
+    YAML.load_file(Rails.root.join("app/lib/accessible_format_request/questions.yaml")).with_indifferent_access
+  end
+
+  def questions
+    @questions ||= question_file["questions"]
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb:15]: again, I went back and forth here as well but agree this is probably the right approach!

File: app/lib/accessible_format_request/questions.yaml [hunk 8]
```diff
@@ -19,10 +20,14 @@ questions:
     type: text
     text: What accessible format do you require?
     caption: Please enter your details here
+    optional: true
+    trigger-key: set-format
+    trigger-value: format-other
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/lib/accessible_format_request/questions.yaml:25]: I don't think there are any plans to seriously change the question structure but useful to have this idea recorded for the future just in case

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        expect(response.body).to have_button("Next")
+      end
+
+      it "passes existing parameters to the next form submit" do
+        expect(response.body).to have_css("form#accessible-format-request[action='#{base_path}?set-format=#{example_param}']")
+      end
+
+      it "uses a GET request for the next form submit" do
+        method = "GET"
+        expect(response.body).to have_css("form#accessible-format-request[method='#{method}']")
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:34]: ah I see. I pulled it out to try and make the important element clearer but if it's had the opposite effect I'll remove it!

File: spec/models/accessible_format_request_spec.rb [hunk 10]
```diff
@@ -42,16 +42,45 @@
...
       expect(subject[:reference]).to eq "accessible-format-request-#{accessible_format_request.object_id}"
     end
 
     it "includes the default reply_to_id" do
       expect(subject[:email_reply_to_id]).to eq "fake-test-accessible-format-request-reply-to-id"
     end
+
+    context "when a standard format type has been selected" do
+      it "includes the format type capitalized with underscores removed" do
+        expect(subject[:personalisation][:format_type]).to eq "Large print"
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply spec/models/accessible_format_request_spec.rb:73]: we pass the presented_format_type so it is human-friendly in the generated email
``` 
 def presented_format_type
    format_type.gsub("-", " ").humanize
 end
  ```

File: app/models/accessible_format_request.rb [hunk 11]
```diff
@@ -16,4 +16,35 @@ def initialize(attributes = {})
...
+      },
+      reference: "accessible-format-request-#{object_id}",
+      email_reply_to_id: reply_to_id,
+    }
+  end
+
+private
+
+  def content_email_address
+    "placeholder"
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/models/accessible_format_request.rb]: this comes in via the content item

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 12]
```diff
@@ -87,6 +87,10 @@ def button_text
     last_question? ? "Submit" : "Next"
   end
 
+  def document_title
+    params[:'document-title'] || "your document"
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb]: this is now coming in via the content_item, and is used for presentation throughout the views and the final email

File: app/lib/accessible_format_request/questions.yaml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    trigger-value: custom
+    inputs:
+      - type: text
+        key: custom-details
+  - group: contact-details
+    title: In order for us to reach you about your request
+    caption: Enter your contact details
+    optional: false
+    inputs:
+      - type: multi-text
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/lib/accessible_format_request/questions.yaml]: this has been updated with help from frontend devs :)

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    "#{contact_govuk_accessible_format_request_path}#{document_ids}?#{permitted_params.to_query}"
+  end
+
+  def permitted_params
+    question_keys = questions.map { |q| q["inputs"].pluck(:key) }.flatten
+    params.permit(*question_keys)
+  end
+
+  def http_method
+    last_question? ? "POST" : "GET"
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb]: this is now outdated as we only use POST requests

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 15]
```diff
@@ -0,0 +1,72 @@
+class Contact::Govuk::AccessibleFormatRequestsController < ContactController
+  helper_method :question_type, :question_title, :question_caption, :question_inputs, :question_options, :question_key, :last_question?, :next_question_number, :next_question_url, :permitted_params, :http_method, :button_text
+
+  def new; end
+
+  def create; end
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb]: thank you, have tidied this up

File: app/lib/accessible_format_request/questions.yaml [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    optional: true
+    trigger-key: format-type
+    trigger-value: custom
+    inputs:
+      - type: text
+        key: custom-details
+  - group: contact-details
+    title: In order for us to reach you about your request
+    caption: Enter your contact details
+    optional: false
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/lib/accessible_format_request/questions.yaml:39]: thank you, have attempted to break this down and re-commit with better discretion and clarity

File: app/models/accessible_format_request.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+  attr_accessor :content_id, :attachment_id, :format_type, :custom_details, :contact_name, :contact_email
+
+  validates :content_id, presence: true
+  validates :attachment_id, presence: true
+  validates :format_type, presence: true
+  validates :contact_name, presence: true
+  validates :contact_email,
+            presence: true,
+            email: { message: "The email address must be valid" },
+            length: { maximum: 1250 }
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/models/accessible_format_request.rb]: Now validating with 
`email: { message: "The email address must be valid" }`

File: app/models/accessible_format_request.rb [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+  def initialize(attributes = {})
+    attributes.each do |key, value|
+      send("#{key}=", value) if respond_to? "#{key}="
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/models/accessible_format_request.rb:17]: this will only attempt to assign an attribute if the attribute actually exists on the model.
It's unlikely to be needed as we also control the model creation in the controller, so happy to remove it if it's more confusing

File: app/views/contact/govuk/accessible_format_requests/form.html.erb [hunk 19]
```diff
@@ -6,19 +6,22 @@
     <%= render "govuk_publishing_components/components/radio", {
       name: question_key,
       heading: question_caption,
-      items: question_options
+      items: question_options,
+      error_message: flash["input_errors"]&.[](question_key)
```
Comment (ChrisBAshton) [inline app/views/contact/govuk/accessible_format_requests/form.html.erb]: I haven't come across this syntax before - specifically the `[](question_key)` bit. What is that doing?

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb [hunk 20]
```diff
@@ -93,6 +93,62 @@
...
+            expect(response.body).to have_css("input[name=question_number][value=#{required_input_question_number + 1}]", visible: false)
+          end
+
+          it "does not set error messages in the flash" do
+            expect(flash["input_errors"]).to eq nil
+          end
+        end
+      end
+
+      context "questions with unrequired inputs" do
```
Comment (ChrisBAshton) [inline spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:195]: Can we make this consistent with the naming in the earlier commits? I.e.

```suggestion
      context "questions with optional inputs" do
```

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb [hunk 21]
```diff
@@ -93,6 +93,62 @@
...
+    context "error messages" do
+      let(:required_inputs) { Hash[required_keys.collect { |v| [v.to_sym, example_param] }] }
+      context "questions with required inputs" do
+        let(:required_input_question) { format_request_questions.detect { |question| question["inputs"][0]["error_message"] } }
+        let(:required_input_question_number) { format_request_questions.find_index(required_input_question) + 1 }
+        describe "when a required input is not entered" do
+          before { post :form, params: base_params.merge({ 'question_number': required_input_question_number + 1, 'previous_question_number': required_input_question_number }) }
+          it "reposts a form requesting the previous question" do
+            expect(response.body).to have_css("input[name=question_number][value=#{required_input_question_number}]", visible: false)
+          end
```
Comment (ChrisBAshton) [inline spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:167]: There's a lot to unpack here! Newlines before each new indentation block (`context`, `before` etc) would help somewhat. I also wonder if some of these `let`s can be moved inline inside the `it` block, as standard variables, which should make this easier to mentally parse.

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 22]
```diff
@@ -6,7 +6,9 @@ class Contact::Govuk::AccessibleFormatRequestsController < ContactController
   before_action :show_previous_with_errors_if_missing_input
   before_action :increment_question_number_if_optional_skipped, only: [:form]
 
-  def form; end
+  def form
+    @content_base_path = content_base_path
```
Comment (ChrisBAshton) [inline app/controllers/contact/govuk/accessible_format_requests_controller.rb:11]: We seem to be using a mixture of class variables (`@content_base_path`) and helper methods (e.g. `question_component`). Can we make `content_base_path` a helper method for consistency, and remove the class variables?

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb [hunk 23]
```diff
@@ -51,6 +51,23 @@
 
   before { stub_publishing_api_has_item(content_item) }
 
+  describe "#unfulilled_request" do
```
Comment (ChrisBAshton) [inline spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:54]:
```suggestion
  describe "#unfulfilled_request" do
```

**Review by ChrisBAshton — COMMENTED** [review]
The switch to exclusively using POST methods is nice 👍

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 24]
```diff
@@ -82,4 +102,28 @@ def display_optional_question?
...
+  def content_base_path
+    content_item["base_path"]
+  end
+
+  def content_attachments
+    content_item.dig("details", "attachments")
+  end
+
+  def requested_attachment
+    @requested_attachment ||= content_attachments.find { |a| a["id"] == params[:attachment_id] }
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb:146]: This one doesn't seem to work, I think it might be that `find ` by implicit id is available on ActiveRecord models - but `content_attachments` is simply an array of hashes?

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 25]
```diff
@@ -82,4 +102,28 @@ def display_optional_question?
...
+  end
+
+  def requested_document_title
+    requested_attachment["title"]
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb:150]: At this stage this is covered by a method in the ApplicationController to log the exception
https://github.com/alphagov/feedback/blob/21a776c007e8b1c515b2f6ac7ab2ccfb310146d1/app/controllers/application_controller.rb#L6

In a later commit, `Add error handling for a failed AccessibleFormatRequest`, we adapt this method to show a custom error page as well as logging the exception

File: app/views/contact/govuk/accessible_format_requests/form.html.erb (hunk 19 above)
Comment (edwardkerry) [reply app/views/contact/govuk/accessible_format_requests/form.html.erb]: It's a rather clunky way of accessing the key in the `FlashHash`, which doesn't seem to allow direct `[question_key]` access at a second level
https://api.rubyonrails.org/classes/ActionDispatch/Flash/FlashHash.html#method-i-5B-5D

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb (hunk 24 above)
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb:146]: Oops, my mistake!

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb (hunk 20 above)
Comment (edwardkerry) [reply spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:195]: Here I've tried to make a distinction between 
`optional questions` which may or may not appear depending on earlier questions vs
`required inputs` which are answers that must be provided. A question could have a mix of required (e.g email address) and non-required (e.g name) inputs

understand this naming may be confusing and/or complex. Perhaps `answers` rather than `inputs`?

Comment (ChrisBAshton) [reply spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:195]: Oh ok. I think what you have or what I've suggested would both work in that case 👍 but perhaps add a comment above the `context` along the lines of what you've just explained, e.g. `# Questions can have a mix of required (e.g email address) and non-required (e.g name) inputs`

File: spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb (hunk 21 above)
Comment (edwardkerry) [reply spec/controllers/contact/govuk/accessible_format_requests_controller_spec.rb:167]: I've attempted to tidy this, it is complicated! Adding line breaks and some comments, but unfortunately the `lets` are already in the last possible point without causing duplication, as they are used in multiple tests

File: app/controllers/contact/govuk/accessible_format_requests_controller.rb [hunk 26]
```diff
@@ -141,4 +142,23 @@ def requested_document_title
...
+    end
+    input_errors = {}
+    previous_input_keys_and_errors.each do |key, error_message|
+      if error_message && (params[key] == "" || params[key].nil?)
+        input_errors[key] = error_message
+      end
+    end
+    if input_errors.any?
+      flash[:input_errors] = input_errors
+      redirect_post(request.original_url, params: permitted_params.to_h.merge(question_number: params[:previous_question_number]))
```
[the remark this answers is not in the document]
Comment (edwardkerry) [reply app/controllers/contact/govuk/accessible_format_requests_controller.rb:176]: It _shouldn't_...
* A user submits question 2 without a required input , which will submit parameters including: `question_number: 3, previous_question_number: 2`
* We detect that an input is missing so add the error message to the flash
* We repost, requesting 2 as the question_number again, including all previously submitted params excluding the previous_question_number

Provided we are always re-requesting the just answered question I think this should be OK, a loop _would_ occur if we reposted to the same question number...

**Review by 1pretz1 — COMMENTED** [review]
All shaping up nicely @edwardkerry, nice work!

Sorry to add another layer of comments. Have left a few small comments but wanted to address the main one here, which is concerning a single action to manage multiple forms. The current approach adds quite a lot of complexity to the flow as we’re maintaining a cross request state (`current_question`), which is a little hard to follow and adds to a bloated controller. I think this would make complete sense for a complicated set of questions, ie a Smart Answer, but for this simple flow, I think we could probably simplify the code.

Could we not simply use different routes / actions for each page? This would remove the need to maintain a state across requests, as we could just direct a user to the correct path depending on what radio button they’ve selected. We could still keep the view code DRY by using a shared template for the common elements.

Following the above approach would also make the [Repost Gem](https://github.com/vergilet/repost) redundant as we could get away with using `render` instead of attempting to redirect a POST request. This approach will add more routes and actions (GET & POST per view), but should simplify the code. We may also be able to get away with using just GET requests and pass the params around in the query string. 

What are your thoughts? This is probably one for @ChrisBAshton too.

**Comment by ChrisBAshton** [comment]
That sounds a good suggestion, @1pretz1, as the use of the 'repost' gem is a bit of a red flag. The proposal is more of a conventional way of doing things 👍 

I would advise attempting that approach in a new branch, Ed, rather than attempt a significant rewrite of this branch. That way, if the new approach turns out not to offer the simplification benefits we hope it will, then we still have the original PR we can tweak and merge. For what it's worth, this PR is close to merging IMO, but I do think it would be worth exploring the other approach first.

**Comment by edwardkerry** [comment]
Superceded by https://github.com/alphagov/feedback/pull/1363
